### PR TITLE
Lite bridge

### PR DESCRIPTION
**Pull request by NPavie**
Pipeline 2 lite build for windows, including a java class to use as bridge for JNI calls

### CODE REVIEW

**Comment by bertfrees** [comment]
@NPavie I've pushed my version of your branch to https://github.com/daisy/pipeline-assembly/compare/lite-bridge. I think it's ready. Try running:

    make.exe dir-word-addin

There are some Ruby dependencies, notably nokogiri and rubyzip (check the Makefile).

There is one issue I hadn't noticed yet, namely that the messages queue contains the new messages of *all* running jobs. But we'll tackle that later.

File: Makefile [hunk 1]
```diff
@@ -175,15 +175,17 @@ target/assembly-$(assembly/VERSION)-word-addin : src/main/jre/OpenJDK11U-jdk_x64
...
-	Zip::File.open("$<") { |zip_file|                \
-		zip_file.each { |f|                          \
-			f_path=File.join("$(dir $@)", f.name);   \
-			FileUtils.mkdir_p(File.dirname(f_path)); \
-			zip_file.extract(f, f_path);             \
-		}                                            \
-	}
+	require 'fileutils';                            	 \
+	require 'zip';                                  	 \
+	if !Dir.exist?("$<"[0..-5]);						 \
```
Comment (bertfrees) [inline Makefile:180]: What is the `[0..-5]` for?

I think there should be a more elegant way to deal with this. Why not delete the directory before the extraction happens? In that case the recipe should not be triggered unless the zip file is newer.

Comment (bertfrees) [reply Makefile]: See my latest commit.

**Comment by egli** [comment]
Do I understand this correctly that this change adds a new dependency namely ruby for everyone who wants to build the assembly?

If so then I'm strictly against it.

From looking at the changes then it looks like some minor convenience functions. For that we certainly do not need an additional dependency. Especially ruby and some gems that have a history of not being very backwards compatible. It is an absolute horror to think that I will have to update some random ruby gems (which will probably fail because my ruby is too old) before I can build the assembly.

**Comment by bertfrees** [comment]
@egli Ruby was added to make the script work on Windows.

**Comment by bertfrees** [comment]
@egli We'll add a check that a certain version of Ruby is required.

**Comment by egli** [comment]
If it only affects the windows build then I do not care.

However if it affects all builds then I'm still totally against it even if you pin the ruby version. There has been a number of times when I tried to build assembly and it failed because I didn't have the right version of some ruby gem installed. 

I implore you you keep the number of dependencies as low as possible. Especially in this case the benefits (at least compared to the bash code you had before) are so little, that it doesn't outweigh the cost you are incurring with an additional build dependency such as ruby.

**Comment by bertfrees** [comment]
It affects all builds because I'm not going to keep a separate Makefile for Windows. I need a scripting language that can be used on all platforms, not only Unix based. I'm replacing the Unix dependency with a new Ruby dependency.

Would Perl be more acceptable?

**Comment by egli** [comment]
> Would Perl be more acceptable?

As stupid as it sounds: yes Perl would be more acceptable. 

As far as I can tell Perl is dead, nothing about it is ever going to change. It is also part of every system pretty much like bash, so not really an additional dependency in my books.

**Comment by bertfrees** [comment]
OK good to know, thanks. I'm still thinking about the best solution to do this, and Perl was another option. And there are some other options. I've started to experiment with pipeline-assembly but eventually want to apply the idea to the super build.

**Comment by bertfrees** [comment]
I am now seriously considering using Java for the Make recipes.

**Comment by egli** [comment]
But I don't really understand, if Nicolas can run Make (on Windows) why can't he run bash. I thought there was WSL (Windows Subsystem for Linux)

**Comment by bertfrees** [comment]
But that's also a dependency, isn't it?

**Comment by egli** [comment]
Python might be another option (for x-platform scripting) as it is pretty ubiquitous, in the sens of installed everywhere. For these kinda things it is probably best to go for the most boring solution.

OTOH, on the train ride home I watched a talk about [babashka tasks](https://www.youtube.com/watch?v=u5ECoR7KT1Y). Looks very interesting, but then you really have an exotic dependency. But maybe for complicated stuff not such a bad idea. Towards the end he has an example where he publishes a release for some clojure linter vs code plugin. Quite involved with make style dependent tasks, etc.

**Comment by egli** [comment]
Yes, of course Python is also a dependency, but it is installed everywhere, comes with batteries included, so probably no gems to install separately that might require a specific version of ruby that doesn't exist on Debian anymore, etc, etc. You get the gist.

**Comment by egli** [comment]
Oh, you mean WSL is a dependency? I thought that was preinstalled on Windows 10 or something?

**Comment by egli** [comment]
> I am now seriously considering using Java for the Make recipes.

I thought people from that world use groovy. If you go that way then it might really be time to look at babashka :-)

**Comment by bertfrees** [comment]
I was thinking Java because Pipeline itself and Maven are already Java. If needed Java libraries can be used by simply dropping the JARs somewhere.

> Oh, you mean WSL is a dependency? I thought that was preinstalled on Windows 10 or something?

I have no idea. To be honest I hadn't heard about WSL yet. It sounds easier to install than MinGW or Cygwin at least.

But I still like the Java idea better. It would be truly portable in the sense that you don't depend on which tools are included in WSL or in your Linux distribution, or which tools you have installed on your MacOS.

**Comment by bertfrees** [comment]
OK, OK, I'll look at babashka.

**Comment by NPavie** [comment]
Hi !
As an alternative for windows, before the makefile was updated, i was keeping on hand the maven commands i needed to directly launch the build.
So an alternative to the makefile could be to provide (in the readme) the list of target and their corresponding maven commands (with the profiles list).
That would be fine by me.

I just have two questions:
- Could the ruby code in the makefile (download and unzip of jres mainly) be replaced by a maven profile ?
- If the goal is to minimize dependency, could it be simpler to keep the makefile for unix user (with curl/unzip for jres) and provide an secondary simple script for windows ?
(like a simple a batch or powershell2 script that map targets to profiles lists and launch maven accordingly)

Switching to a more cross-platform "make" tool is indeed also an option.

**Comment by bertfrees** [comment]
> Could the ruby code in the makefile (download and unzip of jres mainly) be replaced by a maven profile

In theory yes. You could use Ant within Maven (the antrun-plugin), and/or create custom Maven plugins to do stuff. But that just isn't my idea of a good build script. Why write a whole lot of boilerplate code and sacrifice processor time to run stuff from within Maven if you don't need Maven to run it?

This argument applies even more if you keep in mind that my real goal is to make the super-build script platform-independent (not only the assembly build).

I understand the resistance to adding the Ruby dependency though. That issue needs to be addressed in a next iteration. I'll most likely go for Java.

> If the goal is to minimize dependency, could it be simpler to keep the makefile for unix user (with curl/unzip for jres) and provide an secondary simple script for windows ?
(like a simple a batch or powershell2 script that map targets to profiles lists and launch maven accordingly)

No, that is absolutely not an option for me :-). The goal is not to minimize dependencies, the goal is to make it truly cross-platform, and at the same time maintainable. The idea of having to maintain a separate batch script makes me cringe. It is already bad enough that we have the separate launch scripts "pipeline2" and "pipeline2.bat" (by the way the goal is also to replace these with a single cross-platform script, see https://github.com/daisy/pipeline-tasks/issues/120).

**Comment by bertfrees** [comment]
@egli babashka is an interesting project, but as much as I like Clojure, Java seems the better fit for a Java application. babashka-tasks is also interesting, but I'm afraid I'm not ready to part ways with Make.

**Comment by bertfrees** [comment]
Merged in a469728b04.

**Comment by bertfrees** [comment]
Regarding the Ruby dependency: this has been fixed by using Java as the shell language (see https://github.com/daisy/pipeline-assembly/pull/205).

**Pull request closed without merging**